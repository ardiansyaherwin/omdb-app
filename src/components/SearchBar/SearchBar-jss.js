import { fade } from "@material-ui/core/styles";
const styles = (theme) => ({
  grow: {
    flexGrow: 1,
  },
  search: {
    position: "relative",
    borderRadius: theme.shape.borderRadius,
    backgroundColor: fade(theme.palette.common.white, 0.15),
    "&:hover": {
      backgroundColor: fade(theme.palette.common.white, 0.25),
    },
    marginRight: theme.spacing(2),
    marginLeft: 0,
    width: "100%",
  },
  searchIcon: {
    padding: theme.spacing(0, 2),
    height: "100%",
    position: "absolute",
    pointerEvents: "none",
    display: "flex",
    alignItems: "center",
    justifyContent: "center",
  },
  inputRoot: {
    color: "inherit",
    width: "100%",
  },
  inputInput: {
    padding: theme.spacing(1, 1, 1, 0),
    paddingLeft: `calc(1em + ${theme.spacing(4)}px)`,
    transition: theme.transitions.create("width"),
    width: "100%",
  },
  helperText: {
    fontSize: "8px",
    fontWeight: "100",
    color: "white",
    position: "absolute",
    bottom: -14,
    opacity: 0.7,
    letterSpacing: ".75px",
    marginTop: 0,
  },
  progress: {
    position: "absolute",
    top: 2,
    right: 2,
    opacity: 0.7,
  },
  clearIcon: {
    position: "absolute",
    top: 2,
    right: 2,
  },
});

export default styles;
