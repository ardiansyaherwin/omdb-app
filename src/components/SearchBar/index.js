import React, { useState } from "react";
import { PropTypes } from "prop-types";
import { withStyles } from "@material-ui/core/styles";
import {
  AppBar,
  Container,
  Toolbar,
  InputBase,
  FormHelperText,
  CircularProgress,
  IconButton,
} from "@material-ui/core";
import { Search, Clear } from "@material-ui/icons";
import styles from "./SearchBar-jss";

const SearchBar = (props) => {
  const { classes, onSubmit, minimumLength, loading, placeholder, value } = props;
  const [searchBarState, setSearchBarState] = useState({
    error: false,
    value: value,
  });

  function handleKeyPress(e, onSubmit) {
    const value = e.target.value;
    const error = value.length < minimumLength;

    if (e.keyCode === 13 && value.length >= minimumLength) onSubmit(value);

    setSearchBarState({ value, error });
  }

  function handleInputChange(e) {
    const value = e.target.value;
    const error = value.length < minimumLength;
    setSearchBarState({ value, error });
  }

  function handleClear() {
    onSubmit("");
    setSearchBarState({ value: "" });
  }

  return (
    <div className={classes.root}>
      <AppBar position="fixed">
        <Toolbar>
          <Container maxWidth="sm">
            <div className={classes.search}>
              <div className={classes.searchIcon}>
                <Search />
              </div>
              <InputBase
                placeholder={placeholder}
                autoFocus
                classes={{
                  root: classes.inputRoot,
                  input: classes.inputInput,
                }}
                inputProps={{
                  "aria-label": "search",
                  disabled: loading,
                }}
                value={searchBarState.value}
                onChange={handleInputChange}
                onKeyDown={(e) => handleKeyPress(e, onSubmit)}
              />
              {searchBarState.value.length >= minimumLength ? (
                <IconButton
                  aria-label="clear"
                  size="small"
                  className={classes.clearIcon}
                  color="inherit"
                  onClick={handleClear}
                >
                  <Clear />
                </IconButton>
              ) : (
                ""
              )}
              {loading && (
                <CircularProgress
                  className={classes.progress}
                  size={30}
                  color="inherit"
                />
              )}
              {minimumLength && searchBarState.error && (
                <FormHelperText className={classes.helperText}>
                  Minimum {minimumLength} character(s)
                </FormHelperText>
              )}
            </div>
          </Container>
        </Toolbar>
      </AppBar>
    </div>
  );
};

SearchBar.propTypes = {
  classes: PropTypes.object,
  onSubmit: PropTypes.func.isRequired,
  loading: PropTypes.bool,
  minimumLength: PropTypes.number,
  placeholder: PropTypes.string,
  value: PropTypes.string,
};

SearchBar.defaultProps = {
  onSubmit: () => {},
  loading: false,
  minimumLength: 1,
  placeholder: "Search here...",
  value: ''
};

export default withStyles(styles)(SearchBar);
