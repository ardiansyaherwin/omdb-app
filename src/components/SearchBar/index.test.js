import React from 'react'
import { shallow, mount } from 'enzyme'
import SearchBar from './index'
import {
  InputBase,
  CircularProgress,
  IconButton,
  FormHelperText,
} from "@material-ui/core"
import { unwrap } from "@material-ui/core/test-utils"

const SearchBarNaked = unwrap(SearchBar)


const defaultProps = {
  loading: false,
  minimumLength: 1,
  placeholder: "Search here...",
  value: "",
}

const requiredProps = {
  classes: {},
  onSubmit: jest.fn()
}

describe('<SearchBar />', () => {
  
  it("should render as expected", () => {
    const wrapper = shallow(<SearchBar {...requiredProps} {...defaultProps} />)
    expect(wrapper).toMatchSnapshot()
})

  it("should render <CircularProgress /> if defaultProps:loading is TRUE", () => {
    const wrapper = shallow(
      <SearchBarNaked {...requiredProps} {...defaultProps} loading={true} />
    )
    expect(wrapper.find(CircularProgress).exists()).toBeTruthy()
  })

  it("should render close <IconButton /> if props:value.length more than or equal to props:minimumLength", () => {
    const wrapper = shallow(
      <SearchBarNaked
        {...requiredProps}
        value={"abcd"}
        minimumLength={3}
      />
    )
    expect(wrapper.find(IconButton).exists()).toBeTruthy()
  })

  it("should render <FormHelperText /> if props:value.length less than or equal to props:minimumLength", () => {
    const wrapper = shallow(
      <SearchBarNaked
        {...requiredProps}
        value={"ab"}
        minimumLength={3}
      />
    )
    expect(wrapper.find(FormHelperText).exists()).toBeTruthy()
  })

  it("expect to call props:onSubmit if props:minimumLength is less than or equal to props:value", () => {
    const wrapper = mount(
      <SearchBarNaked {...requiredProps} minimumLength={1} value={"ab"} />
    )
    wrapper
      .find(InputBase)
      .at(0)
      .find("input")
      .simulate("keyDown", { keyCode: 13 })

    wrapper.update()
    expect(requiredProps.onSubmit).toHaveBeenCalledTimes(1)
  })
})