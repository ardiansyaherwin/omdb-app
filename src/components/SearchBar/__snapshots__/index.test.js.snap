// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchBar /> should render as expected 1`] = `
<SearchBar
  classes={
    Object {
      "clearIcon": "SearchBar-clearIcon-8",
      "grow": "SearchBar-grow-1",
      "helperText": "SearchBar-helperText-6",
      "inputInput": "SearchBar-inputInput-5",
      "inputRoot": "SearchBar-inputRoot-4",
      "progress": "SearchBar-progress-7",
      "search": "SearchBar-search-2",
      "searchIcon": "SearchBar-searchIcon-3",
    }
  }
  loading={false}
  minimumLength={1}
  onSubmit={[MockFunction]}
  placeholder="Search here..."
  value=""
/>
`;
